app/src/main/java/com/quran/labs/androidquran/model/quran/CoordinatesModel.java

```diff
@@ -46,6 +52,125 @@
 
     return Observable.fromArray(pages)
         .map(page -> new Pair<>(page, database.getVersesBoundsForPage(page)))
+        .map(pair -> new Pair<>(pair.first, normalizeMap(pair.second)))
         .subscribeOn(Schedulers.computation());
   }
+
+  private Map<String, List<AyahBounds>> normalizeMap(Map<String, List<AyahBounds>> original) {
+    Map<String, List<AyahBounds>> normalizedMap = new HashMap<>();
+    final Set<String> keys = original.keySet();
+    for (String key : keys) {
+      List<AyahBounds> normalBounds = original.get(key);
+      normalizedMap.put(key, normalizeAyahBounds(normalBounds));
+    }
+    return normalizedMap;
+  }
+
+  private List<AyahBounds> normalizeAyahBounds(List<AyahBounds> ayahBounds) {
+    int total = ayahBounds.size();
+    if (total < 2) {
+      // only one line, nothing to normalize
+      return ayahBounds;
+    }
+
+    AyahBounds middle = null;
+    for (int i = 1; i < (total - 1); i++) {
+      if (middle == null) {
+        middle = ayahBounds.get(i);
+      } else {
+        middle.engulf(ayahBounds.get(i));
+      }
+    }
+
```

Comment:
this function is quite large, I think it is better to split it into more smaller function.
for example here we can directly
```
return middle != null
        ? normalizeAyahBoundsContainMiddleAyah(ayahBounds, middle)
        : normalizeAyahBoundsWithoutMiddleAyah(ayahBounds);
```